app/src/main/java/org/dhis2/utils/granularsync/SMSSyncProvider.kt

```diff
@@ -1,30 +1,50 @@
 package org.dhis2.utils.granularsync
 
+import android.content.Context
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
 import io.reactivex.Completable
 import io.reactivex.Single
 import io.reactivex.observers.DisposableCompletableObserver
 import io.reactivex.observers.DisposableSingleObserver
-import java.util.Date
 import org.dhis2.R
 import org.dhis2.commons.resources.ResourceManager
 import org.dhis2.usescases.sms.SmsSendingService
 import org.hisp.dhis.android.core.D2
 import org.hisp.dhis.android.core.sms.domain.interactor.SmsSubmitCase
-import org.hisp.dhis.android.core.sms.domain.repository.SmsRepository
 import org.hisp.dhis.android.core.systeminfo.SMSVersion
-import timber.log.Timber
-
-class SMSSyncProvider(
-    private val d2: D2,
-    private val conflictType: SyncStatusDialog.ConflictType,
-    private val recordUid: String,
-    private val dvOrgUnit: String?,
-    private val dvAttrCombo: String?,
-    private val dvPeriodId: String?,
-    private val resourceManager: ResourceManager
-) {
-    private val smsSender: SmsSubmitCase by lazy {
-        d2.smsModule().smsSubmitCase()
+
+interface SMSSyncProvider {
+
+    val d2: D2
+    val conflictType: SyncStatusDialog.ConflictType
+    val recordUid: String
+    val dvOrgUnit: String?
+    val dvAttrCombo: String?
+    val dvPeriodId: String?
+    val resourceManager: ResourceManager
+    val smsSender: SmsSubmitCase
+
+    fun isPlayServicesEnabled(): Boolean
+    fun waitForSMSResponse(
+        context: Context,
+        senderNumber: String,
+        onSuccess: () -> Unit,
+        onFailure: () -> Unit
+    ) {
+        // default behaviour
+    }
+
+    fun unregisterSMSReceiver(requireContext: Context) {
+        // default behaviour
+    }
```

Comment:
instead of having empty methods, why don't we create sealed classes?

Reply:
SMSSyncProvider requires two diferent implementations for each flavor. Therefore, some methods are only relevant in the PlayService one, they need to be implemented empty in the default flavor.